core/sec/security.go

```diff
@@ -20,10 +20,10 @@ type SecureConn interface {
 type SecureTransport interface {
 	// SecureInbound secures an inbound connection.
 	// If p is empty, connections from any peer are accepted.
-	SecureInbound(ctx context.Context, insecure net.Conn, p peer.ID) (SecureConn, error)
+	SecureInbound(ctx context.Context, insecure net.Conn, p peer.ID, muxers []string) (SecureConn, error)
```

Comment:
Have you considered passing the muxers to the constructor of the transport? What pros and cons do you see?

Reply:
Yes, I assume you are referring to the SecurityTransport. It may save us some "nil" filling, especially in the test code, but my thought was that if we pass it to the constructor, the muxers need to be available at Transport construction time, but our config infra does not enforce configure ordering, so that the orders of following are both legitimate: 
                libp2p.Security(tls.ID, tls.New),
                libp2p.Muxer("/mplex/6.7.0", mplex.DefaultTransport),
                libp2p.Muxer("/yamux/1.0.0", yamux.DefaultTransport),

    or 
                libp2p.Muxer("/mplex/6.7.0", mplex.DefaultTransport),
                libp2p.Muxer("/yamux/1.0.0", yamux.DefaultTransport),
                libp2p.Security(tls.ID, tls.New),
                
In the first case, the muxers are not available yet so I see a problem with passing to muxers. Is that concern true or we have other ways of constructing the security transport?

Reply:
There's some magic going on there. The actual construction happens later in https://github.com/libp2p/go-libp2p/blob/17a342c3c70876b9bb6d9b0dc315f9ca6628f483/config/config.go#L165

Reply:
Going through more details of the config module and the reflect based constructors, I believe we can have a way of passing the muxers to the SecurityTransport constructor. That requires adding more code to the relfect based implementation we currently have. Pros of this way is that we may cut some of the verbose changes in the SecureInbound/SecureOutbound functions. Cons of this approach is that It adds to the reflect code which I am hesitate to build up more, because of code safety, readability and debug-ability concerns. But I am open to hear more reasons that we should build up our reflect magic rather than planning to refactor and replace the reflect code in the future.

Reply:
> That requires adding more code to the relfect based implementation we currently have. Pros of this way is that we may cut some of the verbose changes in the SecureInbound/SecureOutbound functions. Cons of this approach is that It adds to the reflect code which I am hesitate to build up more, because of code safety, readability and debug-ability concerns.

I understand your concerns. That reflection magic definitely falls in the category of "(too) clever code". The better way would be to use proper dependency injection. That's a different kind of magic, but at least that magic is maintained by someone else ;) We've been considering making the switch for some time, but the complexity of [fx](https://github.com/uber-go/fx) has scared us away from that so far...

My vote would be for using the reflection logic we currently have. Adding one more type there doesn't make things a lot worse.

Reply:
Alright, that's fair. On implementing the muxer selection with some new reflect magic.